### PR TITLE
unix,win: wait for threads to start

### PR DESCRIPTION
**Pull request by bnoordhuis**
Respin of https://github.com/libuv/libuv/pull/1639 using a semaphore, see https://github.com/libuv/libuv/pull/1660#issuecomment-348481034.

CI: https://ci.nodejs.org/job/libuv-test-commit/616/

Node.js master + commit: https://ci.nodejs.org/job/node-test-commit/14509/

### CODE REVIEW

File: src/threadpool.c [hunk 1]
```diff
@@ -158,10 +172,15 @@ static void init_threads(void) {
...
 
   for (i = 0; i < nthreads; i++)
-    if (uv_thread_create(threads + i, worker, NULL))
+    if (uv_thread_create(threads + i, worker, &ts))
       abort();
 
-  initialized = 1;
+  do {
+    uv_sem_post(&ts.sem);
+    uv_sem_wait(&ts.sem);
```
Comment (davisjam) [inline src/threadpool.c]: This should be "for 1 .. nthreads: uv_sem_wait"

File: src/threadpool.c [hunk 2]
```diff
@@ -50,10 +54,16 @@ static void uv__cancelled(struct uv__work* w) {
  * never holds the global mutex and the loop-local mutex at the same time.
  */
 static void worker(void* arg) {
+  struct uv__thread_start* ts;
   struct uv__work* w;
   QUEUE* q;
 
-  (void) arg;
+  ts = arg;
+  uv_sem_wait(&ts->sem);
```
Comment (davisjam) [inline src/threadpool.c]: Don't wait, just post. The init thread can wait nthreads times, once for each post.

File: src/threadpool.c [hunk 3]
```diff
@@ -29,6 +29,11 @@
 
 #define MAX_THREADPOOL_SIZE 128
 
+struct uv__thread_start {
+  uv_sem_t sem;
+  unsigned int nthreads;
```
Comment (davisjam) [inline src/threadpool.c]: No need for this `nthreads` field.

**Comment by bzoz** [comment]
Node.js CI is failing again

**Comment by bnoordhuis** [comment]
Updated to the simple approach.

Libuv CI: https://ci.nodejs.org/job/libuv-test-commit/624/
Node.js CI: https://ci.nodejs.org/job/node-test-commit/14525/

> Much better, but you introduce about twice as much semaphore overhead as necessary.

There's an interesting backstory here that I should have mentioned (but I was strapped for time yesterday so I didn't.)

I tested the simple approach both with libuv's mach-based emulation and MacOS's native `sem_post()` and `sem_wait()` (which you can turn on in libuv with some artifice) and it doesn't work with the latter - posts sometimes get lost - unless you do the double-post dance.

Since Apple deprecated the `sem_*()` functions there's arguably no point in supporting them.  The simple approach it is then.

**Comment by davisjam** [comment]
> Since Apple deprecated the sem_*()

I would have deprecated them too if posts can get lost!

**Comment by davisjam** [comment]
LGTM

**Comment by bnoordhuis** [comment]
Landed in 8a6d1b32c...c73e73c87.

@davisjam Congratulations on your first Reviewed-By tag!

**Pull request closed without merging**